src/nvim/globals.h

```diff
@@ -224,7 +224,7 @@ EXTERN dict_T vimvardict;                   // Dictionary with v: variables
 EXTERN dict_T globvardict;                  // Dictionary with g: variables
 /// g: value
 #define globvarht globvardict.dv_hashtab
-EXTERN int did_emsg;                        // set by emsg() when the message
+EXTERN bool did_emsg;                       // set by emsg() when the message
```

Comment:
Why did you change this to `bool`???

Reply:
Is it not a bool?

Reply:
No.

Reply:
Why not?

Reply:
Because it can be larger than 1.

Reply:
Search for `did_emsg_before` in `eval.c`.

Reply:
`did_emsg_before` is defined as an `int` but only used as a `bool`

Reply:
It is assigned from `did_emsg` which (from as far as I can tell) is only ever `true` or `false`.

Reply:
In message.c code you deleted in this PR.

Reply:
There's a few stray `int`s in `eval.c` that are derived from `did_emsg` (`save_did_emsg`, `err`, `did_emsg_before`) that probably should be `bool`, too?

Reply:
Ok, I see. I have no idea why I did that. We can revert that specific change?

Is this causing any issues with anything?

Reply:
As I said, they must be `int`, not `bool`.

Reply:
#20580

Reply:
Why? What precisely broke now that they can only take values `0` and `1`?

Reply:
It can cause duplicate error messages in expression parsing, and it is causing a unit test failure in #20574.